File: 2021/day_02.py
def part1(data: str):
    data = data.splitlines()
    h, d = 0, 0
    for line in data:
        instruction, value = line.split()
        value = int(value)
        if instruction[0] == "f":
            h += value
        elif instruction[0] == "d":
            d += value
        else:
            d -= value
    return h * d


def part2(data: str):
    data = data.splitlines()
    h, d, a = 0, 0, 0
    for line in data:
        instruction, value = line.split()
        value = int(value)
        if instruction[0] == "f":
            h += value
            d += value * a
        elif instruction[0] == "d":
            a += value
        else:
            a -= value
    return h * d


if __name__ == "__main__":
    test = True
    test = False
    test_input = """"""
    if test:
        puzzle_input = test_input
    else:
        with open("day_02_input.txt", "r") as input_file:
            puzzle_input = input_file.read().strip()
    print(part1(puzzle_input))
    print(part2(puzzle_input))

File: 2020/day_02.rs
use std::{fs::read_to_string, time::Instant};

pub mod part1 {
    pub fn solve(input: &str) -> usize {
        input
            .lines()
            .filter(|line| {
                let mut parts = line.split(' ');
                let mut range = parts.next().unwrap().split('-');
                let min = range.next().unwrap().parse::<usize>().unwrap();
                let max = range.next().unwrap().parse::<usize>().unwrap();
                let letter = parts.next().unwrap().chars().next().unwrap();
                let password = parts.next().unwrap();
                let count = password.chars().filter(|c| *c == letter).count();
                count >= min && count <= max
            })
            .count()
    }
}

pub mod part2 {
    pub fn solve(input: &str) -> usize {
        input
            .lines()
            .filter(|line| {
                let mut parts = line.split(' ');
                let mut range = parts.next().unwrap().split('-');
                let min = range.next().unwrap().parse::<usize>().unwrap();
                let max = range.next().unwrap().parse::<usize>().unwrap();
                let letter = parts.next().unwrap().chars().next().unwrap();
                let password = parts.next().unwrap();
                (password.chars().nth(min - 1).unwrap() == letter)
                    != (password.chars().nth(max - 1).unwrap() == letter)
            })
            .count()
    }
}

pub fn main(test: bool) {
    let test_input = "1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc"
        .to_owned();
    let puzzle_input = if test {
        test_input
    } else {
        read_to_string("inputs/day_02_input.txt").unwrap()
    };
    let start = Instant::now();
    println!("{}", part1::solve(&puzzle_input));
    println!("Run in {:?}", start.elapsed());
    let start = Instant::now();
    println!("{}", part2::solve(&puzzle_input));
    println!("Run in {:?}", start.elapsed());
}
